spanner/client_test.go

```diff
@@ -2239,3 +2239,169 @@ func TestClient_DoForEachRow_ShouldEndSpanWithQueryError(t *testing.T) {
+			iter.Next()
+			iter.Stop()
+
+			checkRequestsForTags(t, server.TestSpanner, 2, qo)
+			tx.Close()
+		}
+	}
+}
+
+func TestClient_ReadWriteTransaction_Tag(t *testing.T) {
+	t.Parallel()
+
+	server, client, teardown := setupMockedTestServer(t)
+	defer teardown()
+	for _, to := range []TransactionOptions{
+		{},
+		{TransactionTag: "tx-tag-1"},
+	} {
+		for _, qo := range []QueryOptions{
+			{},
+			{RequestTag: "request-tag-1"},
+		} {
+			client.ReadWriteTransactionWithOptions(context.Background(), func(ctx context.Context, tx *ReadWriteTransaction) error {
+				iter := tx.QueryWithOptions(context.Background(), NewStatement(SelectSingerIDAlbumIDAlbumTitleFromAlbums), qo)
+				iter.Next()
+				iter.Stop()
+
+				iter = tx.ReadWithOptions(context.Background(), "FOO", AllKeys(), []string{"BAR"}, &ReadOptions{RequestTag: qo.RequestTag})
+				iter.Next()
+				iter.Stop()
+
+				tx.UpdateWithOptions(context.Background(), NewStatement(UpdateBarSetFoo), qo)
+				tx.BatchUpdateWithOptions(context.Background(), []Statement{
+					NewStatement(UpdateBarSetFoo),
+				}, qo)
+
+				// Check for SQL requests inside the transaction to prevent the check to
+				// drain the commit request from the server.
+				checkRequestsForTags(t, server.TestSpanner, 4, QueryOptions{
+					RequestTag:     qo.RequestTag,
+					transactionTag: to.TransactionTag,
+				})
+				return nil
+			}, to)
+			checkCommitRequestForTags(t, server.TestSpanner, QueryOptions{
+				RequestTag:     "", // Should never be set
+				transactionTag: to.TransactionTag,
+			})
+		}
+	}
+}
+
+func TestClient_StmtBasedReadWriteTransaction_Tag(t *testing.T) {
+	t.Parallel()
+
+	server, client, teardown := setupMockedTestServer(t)
+	defer teardown()
+	for _, to := range []TransactionOptions{
+		{},
+		{TransactionTag: "tx-tag-1"},
+	} {
+		for _, qo := range []QueryOptions{
+			{},
+			{RequestTag: "request-tag-1"},
+		} {
+			tx, _ := NewReadWriteStmtBasedTransactionWithOptions(context.Background(), client, to)
+			iter := tx.QueryWithOptions(context.Background(), NewStatement(SelectSingerIDAlbumIDAlbumTitleFromAlbums), qo)
+			iter.Next()
+			iter.Stop()
+
+			iter = tx.ReadWithOptions(context.Background(), "FOO", AllKeys(), []string{"BAR"}, &ReadOptions{RequestTag: qo.RequestTag})
+			iter.Next()
+			iter.Stop()
+
+			tx.UpdateWithOptions(context.Background(), NewStatement(UpdateBarSetFoo), qo)
+			tx.BatchUpdateWithOptions(context.Background(), []Statement{
+				NewStatement(UpdateBarSetFoo),
+			}, qo)
+			checkRequestsForTags(t, server.TestSpanner, 4, QueryOptions{
+				RequestTag:     qo.RequestTag,
+				transactionTag: to.TransactionTag,
+			})
+
+			tx.Commit(context.Background())
+			checkCommitRequestForTags(t, server.TestSpanner, QueryOptions{
+				RequestTag:     "", // Should never be set
+				transactionTag: to.TransactionTag,
+			})
+		}
+	}
+}
+
+func checkRequestsForTags(t *testing.T, server InMemSpannerServer, reqCount int, qo QueryOptions) {
```

Comment:
Better to rename it as "expectedQueryOptions".

Reply:
I don't think QueryOptions is the right name in this case, as `QueryOptions` also contain other options than tags, and those are not checked here. I think checkFor...RequestOptions would fit better here.